src/Nethermind/Nethermind.Evm/CodeAnalysis/CodeDataAnalyzer.cs

```diff
@@ -0,0 +1,182 @@
+        
+        public CodeDataAnalyzer(byte[] code)
+        {
+            MachineCode = code;
+        }
+        
+        public bool ValidateJump(int destination, bool isSubroutine)
+        {
+            _codeBitmap ??= CodeDataAnalyzerHelper.CreateCodeBitmap(MachineCode);
+     
+            if (destination < 0 || destination >= MachineCode.Length)
+            {
+                return false;
+            }
+
+            if (!CodeDataAnalyzerHelper.IsCodeSegment(_codeBitmap, destination))
+            {
+                return false;
+            }
+            
+            if (isSubroutine)
+            {
+                return MachineCode[destination] == 0x5c;
+            }
+
+            return MachineCode[destination] == 0x5b;
+        }
+    }
+    
+    public static class CodeDataAnalyzerHelper
+    {
+        private const UInt16 Set2BitsMask = 0b1100_0000_0000_0000;
+        private const UInt16 Set3BitsMask = 0b1110_0000_0000_0000;
+        private const UInt16 Set4BitsMask = 0b1111_0000_0000_0000;
+        private const UInt16 Set5BitsMask = 0b1111_1000_0000_0000;
+        private const UInt16 Set6BitsMask = 0b1111_1100_0000_0000;
+        private const UInt16 Set7BitsMask = 0b1111_1110_0000_0000;
+
+        private static readonly byte[] _lookup = new byte[8] { 0x80, 0x40, 0x20, 0x10, 0x8, 0x4, 0x2, 0x1, };
+        
+        /// <summary>
+        /// Collects data locations in code.
+        /// An unset bit means the byte is an opcode, a set bit means it's data.
+        /// </summary>
+        public static byte[] CreateCodeBitmap(byte[] code)
+        {
+            // The bitmap is 4 bytes longer than necessary, in case the code
+            // ends with a PUSH32, the algorithm will push zeroes onto the
+            // bitvector outside the bounds of the actual code.
+            byte[] bitvec = new byte[(code.Length / 8) + 1 + 4];
+            
+            byte push1 = 0x60;
+            byte push32 = 0x7f;
+            
+            for (int pc = 0; pc < code.Length;)
+            {
+                byte op = code[pc];
+                pc++;
+                
+                if (op < push1 || op > push32)
+                {
+                    continue;
+                }
+
+                int numbits = op - push1 + 1;
+                
+                if (numbits >= 8) {
+                    for (; numbits >= 16; numbits -= 16)
+                    {
+                        bitvec.Set16(pc);
+                        pc += 16;
+                    }
+                    for( ; numbits >= 8; numbits -= 8)
+                    {
+                        bitvec.Set8(pc);
+                        pc += 8;
+                    }
+                }
+
+                switch (numbits)
+                {
+                    case 1:
+                        bitvec.Set1(pc);
+                        pc += 1;
+                        break;
+                    case 2:
```

Comment:
could do default:
bitvec.SetN(in ref, numbits)

Reply:
then read SetNBitMasks[numbits] inside and update pc by numbits

Reply:
I'm not get it :)

Reply:
ok, ignore, possibly worse anyway